✨ feat : AI 적성 검사 세션 기반 이력 저장 및 조회 기능 구현

sooyeoneo opened the pull request:
## ☀️ 작업 목적

* 적성 AI 검사를 **세션 단위로 저장 및 관리**하여 데이터 일관성과 추적성 강화
* String → Enum(AptitudeType) 통일로 **타입 안정성 확보** 및 서비스 전반의 일관성 유지
* DTO/Service 책임 분리로 **가독성과 유지보수성** 향상
* Redis TTL(30분) 세션 타임아웃을 적용하여 **사용자 경험과 리소스 효율성 개선**
* 완료된 검사 이력/세션별 이력 조회 API 제공

---

## 📄 주요 작업 내역

### 1. AptitudeTestHistory Entity 개선 

* `String aptitudeType` → `AptitudeType analyzedAptitudeType` (Enum 통일)
* 세션 ID, 신뢰도 점수, 완료 여부 필드 추가
* UserAptitude와 동일 Enum 사용으로 일관성 확보
* #55 #151 #164

### 2. AptitudeTestHistoryRepository 정리 

* `findByMemberOrderByStepAsc()` 제거 (세션 기반 대체)
* `markSessionAsCompleted()` 쿼리 메소드 추가
* 각 메소드 용도 주석화 → 유지보수성 강화
* #65 #164

### 3. AptitudeService 세션 기반 이력 관리 

* `generateSessionId()`: `APT-{userId}-{timestamp}` 형식으로 세션 생성
* `parseAptitudeType()`: String → Enum 안전 변환
* `convertToDto()`: Entity → DTO 변환 로직 이동
* 테스트 완료 시 세션 전체를 일괄 완료 처리
* 세션별/완료된 이력 조회 메소드 구현
* #65 #151 #164

### 4. AptitudeCacheService Redis 세션 관리 

* `saveSessionId()`, `getSessionId()`: 세션 ID 저장 및 조회
* `deleteAllTestData()`: 세션 ID 포함 전체 데이터 정리
* 세션 기반 테스트 데이터 격리로 안정성 강화
* #65 #151 #164

### 5. Controller 및 API 추가 

* `GET /ai/aptitude/history`: 완료된 적성 검사 이력 조회 API
* `GET /ai/aptitude/history/session`: 세션별 이력 조회 API
* `AptitudeTestHistoryResponseDto.from()` 제거 → DTO 순수성 확보
* #65 #151 #164

### 6. 세션 타임아웃 설정 

* Redis TTL 30분 적용
* 답변 제출 시마다 TTL 갱신 → 유휴 세션 유지
* 30분 이상 미활동 시 자동 만료 처리
* #65 #151 #164

### 7. 테스트 케이스 및 HTTP 요청 파일 

* `ai-aptitude.http`에 이력 조회 API 테스트 추가
* 테스트 플로우:

  1. 적성 테스트 5단계 완료
  2. 완료된 전체 이력 조회
  3. 반환된 세션 ID로 특정 세션 조회

* #65 #151 #164

---

## 📷 테스트 결과 캡처

1. **전체 이력 조회 API**
<img width="1792" height="1137" alt="적성 검사 이력 조회 테스트 2025-09-06 13 48 36" src="https://github.com/user-attachments/assets/d2aff207-b898-4e2a-b0a2-3975f9fb8d82" />


<img width="1792" height="1137" alt="적성 검사 이력 조회 테스트 2025-09-06 13 49 00" src="https://github.com/user-attachments/assets/7a182c1f-c384-40cd-b449-b9ff5a3d35ec" />


2. **세션별 이력 조회 API**
<img width="1792" height="1137" alt="적성 검사 이력 조회 테스트 2025-09-06 13 50 41" src="https://github.com/user-attachments/assets/2148b5f0-5082-482f-b038-a3eb4e30d47f" />


<img width="1792" height="1137" alt="적성 검사 이력 조회 테스트 2025-09-06 13 51 21" src="https://github.com/user-attachments/assets/052962c9-bdd3-44be-a27a-4b5ac4a65167" />


3. **Redis 로그 확인 (세션별 점수/답변 저장 및 정리)**
<img width="1789" height="1127" alt="적성 검사 이력 조회 테스트 2025-09-06 13 49 32" src="https://github.com/user-attachments/assets/9b4c2cd0-c331-4ab9-9b86-2de15f543301" />


---

## ♻️ 개선 효과

* 세션 기반 이력 추적 가능
* Enum 통일로 **타입 안정성 강화**
* TTL 관리로 **사용자 경험 개선 + 서버 리소스 절약**
* DTO 순수성 유지 → Service 책임 분리로 **코드 품질 향상**
* **중복/불필요 코드 제거**로 가독성 향상

---

## 📌 참고 및 TODO

- 세션 TTL(30분) → 실제 운영에서는 사용자 행동 기반 갱신 로직 재검토
- 프론트엔드 연동 시 UX 개선 (결과 확정 버튼 분리, 프로그램 추천 API와 연계)

---

**리뷰 및 피드백 환영합니다! 🤗**

sooyeoneo committed:
✨ feat : 적성 검사 Entity 및 Enum 구현

- UserAptitude Entity 구현
  - 회원별 적성 정보 저장 (1:1 관계)
  - 테스트 횟수 추적 (최대 5회 제한)
  - BaseEntity 상속으로 생성/수정 시간 자동 관리

- AptitudeTestHistory Entity 구현
  - 적성 검사 이력 저장 (N:1 관계)
  - 질문-답변-AI응답 기록

- AptitudeType Enum 구현
  - 5가지 적성 타입 정의 (자연친화, 역사문화, 예술창작, 커뮤니티, 디지털기술)
  - 각 타입별 제목과 설명 포함

#65

sooyeoneo committed:
✨ feat : 적성 검사 Repository 구현

- UserAptitudeRepository
  - 회원별 적성 조회 (findByMember)
  - 적성 존재 여부 확인 (existsByMember)

- AptitudeTestHistoryRepository
  - 회원별 검사 이력 조회 (오름차순 정렬)
  - 회원별 이력 전체 삭제

#65

sooyeoneo committed:
✨ feat : 적성 검사 DTO 구현

- RequestDto (요청)
  - AptitudeAnswerRequestDto: 사용자 답변 제출

- ResponseDto (응답)
  - AptitudeQuestionResponseDto: AI 질문 정보를 사용자에게 답변 응답
  - AptitudeTestStartResponseDto: 검사 시작 응답
  - AptitudeTextProgressResponseDto: 진행 상태 응답
  - AptitudeTestResultResponseDto: 최종 결과 응답
  - CanRetakeTestResponseDto: 재검사 가능 여부 응답

* 모든 ResponseDto는 불변 객체로 구현 (생성자 사용)

#64 #65

sooyeoneo committed:
✨ feat : 적성 검사 Service 구현

- AptitudeAiService
  - 5개 질문 Pool 정적 관리
  - AI 프롬프트 기반 답변 분석
  - 적성 타입별 점수 계산 및 최종 적성 결정
  - BaseAiService 상속으로 재시도 및 예외처리

- AptitudeService
  - 검사 시작/답변 제출/결과 저장 플로우
  - Redis 기반 검사 진행 상태 관리 (24시간 TTL)
  - 최대 5회 재검사 제한 로직
  - 수동 적성 선택 기능
  - CustomException 기반 예외처리

* @Transactional 메서드별 적용

#64 #65 #66

sooyeoneo committed:
✨ feat : 적성 검사 Controller 구현

- API 엔드포인트
  - POST /api/ai/aptitude/test/start: 검사 시작
  - POST /api/ai/aptitude/test/answer: 답변 제출
  - POST /api/ai/aptitude/select: 수동 적성 선택
  - GET /api/ai/aptitude/can-retake: 재검사 가능 여부

- Spring Security 적용
  - @PreAuthorize(\"hasRole('USER')\") 일반회원 권한 체크
  - MemberDetails 통한 회원 정보 획득

- CommonResponseDto 표준 응답 형식 적용

#63 #66

sooyeoneo committed:
🔧 add : Redis 범용 aiAptitudeRedisTemplate 빈 추가

- RedisTemplate<String, String> 빈 추가
- @Qualifier("aiAptitudeRedisTemplate") 빈에 추가 구분자를 붙여 명시적으로 지정
- 적성 검사 진행 상태 관리용
- 기존 blacklist, emailVerified 템플릿과 별도 관리

#64

sooyeoneo committed:
✅ test : 적성 검사 HTTP 테스트 및 AI 프롬프트 추가

- ai-aptitude.http
  - 적성 검사 시작/답변/선택/재검사 API 테스트
  - Authorization Bearer 토큰 기반

- aptitude-test.txt
  - 적성 분석 AI 프롬프트 추가

#67

sooyeoneo committed:
💡 comment : AiTestController 주석 처리 (추후 삭제 예정)

sooyeoneo committed:
💡 comment : 필요 없는 주석 삭제

sooyeoneo committed:
🔥 remove : @PreAuthorize("hasRole('USER')") 제거

sooyeoneo committed:
🔧 add : NOT_FOUND_TEST_PROGRESS, TOO_MANY_REQUESTS ErrorCode에 추가

sooyeoneo committed:
🔧 add : 적성 검사 AI 응답 분석에 정규식 패턴 매칭 로직 추가

- analyzeHistories() 메서드에 Pattern/Matcher 기반 분석 로직 적용
- 정규식으로 모든 적성 타입(name + title) OR 연결 패턴 생성
- AI 응답 내 복수 적성 키워드 추출 및 점수 계산 지원
- 향후 HashMap 최적화를 위한 TODO 주석 추가

Refs: #65

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java
#	src/main/java/com/salayo/locallifebackend/global/config/RedisConfig.java
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
🔧 add : @PreAuthorize("hasRole('USER')") 추가

- AptitudeController에 @PreAuthorize("hasRole('USER')") 일반 회원 권한을 다시 추가

#55

sooyeoneo committed:
🔧 add : @RequireArgsConstructor 삭제하고, 생성자 직접 주입

- AptitudeController에 @RequireArgsConstructor 삭제하고, 생성자 직접 주입(Constructor injection)으로 의존성을 직접 전달

#55

sooyeoneo committed:
🔥 remove : 사용하지 않는 import 문 제거

sooyeoneo committed:
♻️ refactor : 적성 검사 캐시 로직 분리 및 CacheKeyPrefix로 통합 관리

- AptitudeCacheService 생성하여 Redis 캐시 로직 캡슐화
  - ReviewCacheService와 동일한 아키텍처 패턴 적용
  - 테스트 진행 상태 저장/조회/삭제 메서드 구현

- CacheKeyPrefix에 적성 검사 관련 모든 상수 통합
  - APTITUDE_MAX_TEST_COUNT: 최대 테스트 횟수 (5회)
  - APTITUDE_TOTAL_QUESTIONS: 전체 질문 개수 (5개)
  - APTITUDE_TEST: Redis 키 프리픽스 ("aptitude:test:")
  - APTITUDE_TEST_TTL_HOURS: 캐시 TTL (24시간)

- AptitudeService 리팩토링
  - 캐시 로직을 AptitudeCacheService로 위임
  - CacheKeyPrefix 상수 참조로 중앙 집중식 관리로 유지보수성 향상
  - AptitudeService와 AptitudeCacheService에서 CacheKeyPrefix 참조

#55 #65

sooyeoneo committed:
💡 comment : 주석 수정

sooyeoneo committed:
🐛 fix : 적성 검사 횟수 제한 에러코드 수정

- APTITUDE_TEST_LIMIT_EXCEEDED 에러코드 추가 (400 BAD_REQUEST)
- TOO_MANY_REQUESTS(429)는 Rate Limiting 전용으로 복원
- 비즈니스 로직 제약사항에 맞는 적절한 상태 코드 사용

#55 #65 #67

sooyeoneo committed:
📝 docs : 적성 검사 API Swagger 문서화 추가

- Controller 레벨
  - @Tag 추가: "Aptitude" 그룹화 및 설명
  - @Operation: 각 API 엔드포인트별 요약 및 상세 설명
  - @ApiResponses: HTTP 상태 코드별 응답 설명 (200, 201, 400, 401, 404)
  - @Parameter: 요청 파라미터 설명 및 예시값

- DTO 레벨
  - 모든 Request/Response DTO에 @Schema 추가
  - 필드별 설명, 예시값, 최소/최대값, 필수 여부 명시
  - AptitudeQuestionResponseDto: 질문 정보 및 예시 답변
  - AptitudeAnswerRequestDto: 답변 요청 필드 제약사항
  - CanRetakeTestResponseDto: 재검사 관련 정보

- Enum 레벨
  - AptitudeType에 @Schema 추가
  - 각 적성 타입별 상세 설명 포함

Swagger UI에서 적성 검사 API의 상세 문서 확인 가능

#55 #64 #65 #66

sooyeoneo committed:
🔧 add : @Valid 추가

sooyeoneo committed:
✏️ typos : 코드 줄 정리

sooyeoneo committed:
🔧 add : ResponseDto에 @Builder를 생성자 레벨에 추가

sooyeoneo committed:
♻️ refactor : 적성 검사 횟수 제한 로직 개선 - 온보딩/마이페이지 구분

- UserAptitude 엔티티 개선
  - mypageTestCount 필드 추가 (마이페이지 검사 횟수)
  - isOnboardingCompleted 필드 추가 (온보딩 완료 여부)
  - 온보딩/마이페이지별 업데이트 메서드 분리

- 적성 검사 제한 로직 변경
  - 온보딩 첫 검사는 횟수 제한 없음
  - 마이페이지에서만 최대 5회 제한 적용
  - 총 6회 검사 가능 (온보딩 1회 + 마이페이지 5회)

- API 문서 및 에러 메시지 업데이트
  - 온보딩과 마이페이지 검사 구분 명시
  - APTITUDE_TEST_LIMIT_EXCEEDED 메시지 수정

#55 #66

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/controller/AptitudeController.java
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/dto/CanRetakeTestResponseDto.java
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
♻️ refactor : AptitudeAiService 구조 개선 및 캐시 활용 강화

- AptitudeCacheService에 답변 저장 및 적성 점수 관리 기능 추가
  - saveAnswer(): 단계별 답변 Redis 저장
  - updateAptitudeScore(): 적성별 점수 누적 관리
  - getAllAptitudeScores(): 전체 점수 조회
  - deleteAllTestData(): 테스트 데이터 정리

- AptitudeAiService 단순화 및 개선
  - AI 호출 실패 시 fallback 키워드 분석
  - 로그 레벨 log.warn 설정

- AptitudeService Redis 활용 개선
  - AI 분석 결과를 Redis에 실시간 저장
  - DB 텍스트 파싱 대신 Redis 점수 직접 활용
  - 테스트 완료 후 자동 데이터 정리

#55 #65 #67 #151

sooyeoneo committed:
♻️ refactor : AI 적성 검사 테스트 환경 및 데이터 정리 로직 개선

- HTTP 테스트 파일 개선
  - 동적 변수 처리를 위한 응답 구조 추가(aptitudeQuestionResponseDto 경로 수정)
  - AI 적성 검사 전체 플로우, 개별 단계별, 수동 선택 테스트 섹션 구분
  - 콘솔 로그 추가로 테스트 진행 상황 추적 가능

- Redis 데이터 정리 로직 강화
  - deleteAllTestData() 메소드에 상세 로그 추가
  - 각 데이터 타입별 삭제 결과 확인 (진행상태, 답변, 점수)
  - 테스트 시작 시 기존 데이터 정리로 중복 방지

- 디버깅 로그 추가
  - Redis 데이터 정리 시작/완료 로그
  - 답변 저장 및 점수 업데이트 과정 추적

#55 #65 #67 #151

sooyeoneo committed:
🏗 chore : AI 적성 검사 세션 타임아웃 30분 설정

#55 #64

sooyeoneo committed:
✨ feat : AI 적성 분석 JSON 파싱 구조화 및 신뢰도 기반 점수 적용

- AiAptitudeAnalysisResponseDto 생성 (적성 타입, 신뢰도, 이유, 핵심요소)
- AiQuestionGenerationResponseDto 생성 (추후 동적 질문 생성용)
- AptitudeAiService에서 구조화된 객체 반환 및 JSON 응답 프롬프트 개선
- 신뢰도 점수(0.7 기준)에 따른 가중치 적용 (높음: 2점, 낮음: 1점)
- Fallback 분석도 동일한 객체 구조로 통일
- 테스트용 로그 추가 (JSON 파싱, 가중치 적용 확인)

#55 #64 #67 #151

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
♻️ refactor : 코드 리뷰 반영 - AI 적성 분석 안정성 및 일관성 개선

[변경 내용]
1. AiAptitudeAnalysisResponseDto 개선
   - String[] keyFactors → List<String> keyFactors 변경
   - 방어적 복사(Defensive Copy) 적용으로 null 안전성 향상
   - 프로젝트 내 다른 DTO들과 컬렉션 사용 통일

2. AptitudeService 안전성 강화
   - toUpperCase() 적용으로 소문자 입력 대응
   - AI 응답의 대소문자 구분 없이 안전하게 처리

3. AptitudeAiService fallback 로직 개선
   - List.of() 사용으로 일관성 유지
   - Fallback 시에도 동일한 자료구조 사용

[개선 효과]
- 프로젝트 전체 일관성 확보
- null 및 대소문자 안전성 향상
- Spring/Jackson 표준 준수
- 유지보수성 개선

Resolves: #158
Co-authored-by: jiyoon0000 <rlawldbs9128@gmail.com>

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
✨ feat : AptitudeTestHistory Entity 타입 안정성 개선

- String aptitudeType → AptitudeType analyzedAptitudeType (Enum 타입 통일)
- 세션 ID, 신뢰도 점수, 완료 여부 필드 추가
- UserAptitude와 동일한 Enum 타입 사용으로 일관성 확보

Resolves: #55 #151 #164

sooyeoneo committed:
♻️ refactor : AptitudeTestHistoryRepository 메소드 정리

- findByMemberOrderByStepAsc() 제거 (세션 기반으로 대체)
- 각 메소드 용도 명확히 하는 주석 추가
- markSessionAsCompleted() 쿼리 메소드로 일괄 완료 처리
- 추후 사용 예정 메소드 용도 명시

Resolves: #65 #164

sooyeoneo committed:
✨ feat : AptitudeService 세션 기반 이력 관리 구현

- generateSessionId(): APT-{userId}-{timestamp} 형식 세션 생성
- parseAptitudeType(): String → Enum 안전 변환 헬퍼 메소드
- convertToDto(): Entity → DTO 변환 로직 Service로 이동
- 테스트 완료 시 세션 일괄 완료 처리
- 세션별/완료된 이력 조회 메소드 구현

Resolves: #65 #151 #164

sooyeoneo committed:
✨ feat : AptitudeCacheService Redis 세션 관리 기능 추가

- saveSessionId(), getSessionId(): 세션 ID 저장/조회
- deleteAllTestData(): 세션 ID 포함 전체 데이터 정리
- 세션 기반 테스트 데이터 격리

Resolves: #65 #151 #164

sooyeoneo committed:
✨ feat : 적성 검사 이력 조회 API 및 DTO 개선

- GET /ai/aptitude/history: 완료된 이력 조회 API
- GET /ai/aptitude/history/session: 세션별 이력 조회 API
- AptitudeTestHistoryResponseDto from() 메소드 제거
- DTO 순수성 유지 (데이터만 담당, 로직은 Service)

Resolves: #65 #151 #164

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
✅ test : ai-aptitude.http에 이력 조회 API 테스트 추가

[추가 내용]
- GET /ai/aptitude/history: 완료된 적성 검사 이력 조회
- GET /ai/aptitude/history/session: 특정 세션 ID별 이력 조회
- 세션 ID 예시 및 사용 가이드 콘솔 출력

[Test Flow]
1. 적성 테스트 완료 (5단계)
2. 완료된 이력 전체 조회
3. 반환된 세션 ID로 특정 세션 조회

Resolves: #65 #151 #164

sooyeoneo committed:
✨ feat : 적성 검사 세션 30분 타임아웃 구현

- Redis 세션 TTL 30분 설정
- 답변 제출 시마다 세션 갱신
- 30분 이상 미활동 시 자동 만료
- 세션 재생성 로직 및 로그 추가

[개선 효과]
- 사용자가 잠시 자리를 비워도 테스트 유지
- 방치된 세션 자동 정리로 메모리 효율화

Resolves: #65 #151 #164

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/dto/AptitudeTestHistoryResponseDto.java:
```diff
@@ -0,0 +1,38 @@
+    private final String questionText;
+    private final String userResponse;
+    private final String aiResponse;
+    private final String aptitudeType;
```
AptitudeType을 enum으로 정의해두셨던 걸로 기억하는데, Dto에서는 String 으로 응답하도록 하는 이유가 있을까요?

sooyeoneo replied:
다른 dto에 수정하면서 다 수정했다고 생각했는데, 아직 남아있었네요...
정신없이 String으로 쓰다가 enum의 존재를 누락한 저의 불찰입니다...

```
private final aptitudeType aptitudeType;
```

로 바로 수정하였습니다 🥹

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java:
```diff
@@ -239,6 +280,53 @@ public CanRetakeTestResponseDto canRetakeTest(Long memberId) {
+		List<AptitudeTestHistory> histories = aptitudeTestHistoryRepository.findBySessionIdOrderByStepAsc(sessionId);
+		
+		// 본인의 이력인지 확인
+		if (!histories.isEmpty() && !histories.get(0).getMember().getId().equals(memberId)) {
```
현재 getTestHistoryBySession에서 세션이 아예 존재하지 않는 경우 `histories.isEmpty()`일 때 그대로 빈 리스트로 반환되고 있는데, 잘못되거나 만료된 세션 ID라면 `404 Session Not Found`로 처리하는게 세션 없음과 이력 없음이 구분이 되어 조금 더 직관적일 것 같아요!

```
if (histories.isEmpty()) {
    throw new CustomException(ErrorCode.SESSION_NOT_FOUND);
}
```

만약, 수연님의 의도가 세션은 존재하지만 아직 답변 기록이 저장되지 않은 상태를 표현한거라면 빈 리스트 반환하는 게 맞을 것 같습니다!

sooyeoneo replied:
코멘트 감사합니다! 😊

말씀해주신 대로 `getTestHistoryBySession()` 메서드에서 세션이 존재하지 않거나 만료되었을 때, 
빈 리스트를 반환하는 것은 직관적이지 않은 것 같습니다.

`histories.isEmpty()`를 확인하여 세션이 유효하지 않은 경우 명확하게 `SESSION_NOT_FOUND` 예외를 발생시키도록 로직을 개선했습니다. 또한, 이어서 세션이 존재하더라도 소유권이 없는 경우 `FORBIDDEN_ACCESS` 예외를 던지도록 로직을 분리하여 예외 처리의 명확성을 높였습니다.

덕분에 훨씬 더 명확해진 것 같아요, 지윤님 감사합니다! 💗

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java:
```diff
@@ -118,24 +142,32 @@ public AptitudeTextProgressResponseDto submitAnswer(Long memberId, AptitudeAnswe
+			.analyzedAptitudeType(parseAptitudeType(aiAnalysis.getAptitudeType()))
+			.confidenceScore(aiAnalysis.getConfidenceScore())
+			.sessionId(sessionId)
+			.isCompleted(false)
```
이력이 `isCompleted(false)`로 저장되고 마지막 단계에서만 `markSessionAsCompleted()`로 호출하는 것 같아요!

제 생각에 이력 저장 시 기본 값이 `isCompleted = false`일 경우 사용자가 테스트를 중도 포기하거나 세션이 만료된 경우에는 미완료 이력이 DB에 남을 수도 있을 것 같아요!

그래서 추후에 세션 만료 시 배치나 스케줄러를 사용하여 미완료 이력을 정리하는 로직도 있으면 어떨까요?
아니면 테스트 완료 시점에만 이력을 영구 저장하고 중간 단계는 redis 캐시에 넣는 방법도 있을 것 같아요!

sooyeoneo replied:
넵 맞습니다! 
말씀하신 대로 현재 로직은 미완료된 테스트 이력이 DB에 남게 됩니다.

사실 이 부분은 의도된 설계였는데요, 

사용자가 테스트를 중단하더라도 **30분 내에 이어하기** 기능을 제공하기 위함이었습니다. 
(현재 `SESSION_TIMEOUT_MINUTES = 30`으로 설정) 
`isCompleted = false`로 저장하여 미완료/완료 상태를 구분하고 있습니다.
(#166 PR 이후, 부분 저장 및 이어하기 기능 PR을 올리려고 준비중입니다.)

---

**개선 내용 PR**
지윤님 의견대로 불필요한 데이터가 쌓이는 문제를 해결하고자 
다음 PR에서 미완료 이력을 정리하는 로직을 추가할 예정입니다.

1. `expiresAt` 필드 추가: 세션 만료 시간(30분)에 맞춰 DB 데이터도 정리
2. 스케줄러 구현: 30분 이상 지난 미완료 세션들을 주기적으로 삭제

---

**Redis 캐시 방식을 선택하지 않은 이유**

Redis에만 저장할 경우 서버 재시작이나 장애 상황에서 데이터 손실 위험이 있어, 
30분 내라도 안정적인 이어하기 기능 제공이 어렵다고 판단했습니다.

현재 PR은 기능 동작에 문제가 없으므로 Merge 후, 미완료 이력 정리 로직은 별도 이슈로 관리하겠습니다.

감사합니다! 😍

---

**+ 추가로 고민인 부분이에요💭**

TTL을 30분으로 할 지, 24시간으로 할지.. 아니면 다른 시간 단위로 할지 고민입니다.. 1시간..?

* 현재 30분의 장단점
  * 장점
     * DB 부담 적음 (빠른 정리)
     * 대부분 테스트는 10-15분 내 완료
  * 단점
     * 잠시 자리 비웠다가 돌아오면 초기화 

30분으로 하면 자주 정리해줘야 해서 스케줄러 꼭 있어야할 것 같네요.. 
"Redis야.. 괜찮니..?😇"

jiyoon0000 replied:
흠 그래서 TTL 시간 관련해서 조금 고민해봤는데요!!

제 생각에는 온보딩 기능이 완료된 후에 사용자 입장에서 테스트를 해보고 예를 들어 테스트가 10분 내외다 하면 30-40분, 15-20분 정도일 경우 1시간 정도로 잡는게 적당할 것 같아요!

DB에 부하를 크게 주지 않으면서도 테스트를 이어서 할 수 있도록 시간 계산하는 게 좋을 것 같아요~! 나중에 다시 같이 고민해봐용😘

sooyeoneo committed:
🐛 fix : AptitudeTestHistoryResponseDto의 aptitudeType의 타입 String → Enum 수정

Resolves: #164 #166

sooyeoneo committed:
🐛 fix : 유효하지 않은 ai-aptitude의 test session에 대한 명시적 예외 처리 추가

[변경 내용]
- 제공된 세션 ID로 이력 목록을 조회했을 때 비어 있는 경우, AptitudeService.getTestHistoryBySession 메서드가 SESSION_NOT_FOUND 예외를 발생시키도록 수정
- 세션 이력이 존재하는 경우에만 소유권 확인하도록 로직을 수정하여, 잠재적인 IndexOutOfBoundsException 발생 방지

[개선 효과]
- 유효하지 않거나 만료된 세션과 이력이 없는 유효한 세션을 구분하여, 존재하지 않는 세션에 접근하려는 사용자에게 더 명확하고 직관적인 오류 응답을 제공

Resolves: #164 #166
